packages/tre/src/plugins/r2/gateway.ts

```diff
@@ -112,28 +191,131 @@ function sqlStmts(db: TypedDatabase) {
+    Pick<MultipartPartRow, "upload_id" | "part_number">,
+    Pick<MultipartPartRow, "blob_id">
+  >(
+    // For getting part number's previous blob ID to garbage collect
+    "SELECT blob_id FROM _mf_multipart_parts WHERE upload_id = :upload_id AND part_number = :part_number"
+  );
+  const stmtPutPart = db.prepare<Omit<MultipartPartRow, "object_key">>(
+    // For recording metadata when uploading parts
+    `INSERT OR REPLACE INTO _mf_multipart_parts (upload_id, part_number, blob_id, size, etag, checksum_md5)
+    VALUES (:upload_id, :part_number, :blob_id, :size, :etag, :checksum_md5)`
+  );
+  const stmtLinkPart = db.prepare<
+    Pick<MultipartPartRow, "upload_id" | "part_number" | "object_key">
+  >(
+    // For linking parts with an object when completing uploads
+    `UPDATE _mf_multipart_parts SET object_key = :object_key
+    WHERE upload_id = :upload_id AND part_number = :part_number`
+  );
+  const stmtDeletePartsByUploadId = db.prepare<
+    Pick<MultipartPartRow, "upload_id">,
+    Pick<MultipartPartRow, "blob_id">
+  >(
+    // For deleting parts when aborting uploads
+    "DELETE FROM _mf_multipart_parts WHERE upload_id = :upload_id RETURNING blob_id"
+  );
+  const stmtDeleteUnlinkedPartsByUploadId = db.prepare<
+    Pick<MultipartPartRow, "upload_id">,
+    Pick<MultipartPartRow, "blob_id">
+  >(
+    // For deleting unused parts when completing uploads
+    "DELETE FROM _mf_multipart_parts WHERE upload_id = :upload_id AND object_key IS NULL RETURNING blob_id"
+  );
+  const stmtDeletePartsByKey = db.prepare<
+    Pick<MultipartPartRow, "object_key">,
+    Pick<MultipartPartRow, "blob_id">
+  >(
+    // For deleting dangling parts when overwriting an existing key
+    "DELETE FROM _mf_multipart_parts WHERE object_key = :object_key RETURNING blob_id"
+  );
+  const stmtListPartsByUploadId = db.prepare<
+    Pick<MultipartPartRow, "upload_id">,
+    Omit<MultipartPartRow, "blob_id">
+  >(
+    // For getting part metadata when completing uploads
+    `SELECT upload_id, part_number, blob_id, size, etag, checksum_md5, object_key
+    FROM _mf_multipart_parts WHERE upload_id = :upload_id`
+  );
+  const stmtListPartsByKey = db.prepare<
+    Pick<MultipartPartRow, "object_key">,
+    Pick<MultipartPartRow, "blob_id" | "size">
+  >(
+    // For getting part metadata when getting values, size included for range
+    // requests, so we only need to read blobs containing the required data
+    "SELECT blob_id, size FROM _mf_multipart_parts WHERE object_key = :object_key ORDER BY part_number"
+  );
+
   return {
-    getByKey: db.prepare<Pick<ObjectRow, "key">, ObjectRow>(`
-      SELECT key, blob_id, version, size, etag, uploaded, checksums, http_metadata, custom_metadata
-      FROM _mf_objects WHERE key = :key
-    `),
+    getByKey: stmtGetByKey,
+    getPartsByKey: db.transaction((key: string) => {
+      const row = stmtGetByKey.get({ key });
+      if (row === undefined) return;
+      if (row.blob_id === null) {
```

Comment:
A comment to say that (I think) having a blob_id indicates it is not multi-part?

Reply:
Or more accurately having `null` as blob_id makes it multipart.